### PR TITLE
Implement DeclarePositionRule

### PR DESCRIPTION
**Pull request by staabm**
closes https://github.com/phpstan/phpstan/issues/6195

### CODE REVIEW

**Comment by mad-briller** [comment]
does this ignore `declare` used for [purposes other than](https://phpstan.org/r/afae2682-8ec3-4f74-b555-92daf62bad46) `strict_types`?

**Comment by staabm** [comment]
good catch - fixed thanks

**Comment by ondrejmirtes** [comment]
Please name this accordingly "DeclareStrictTypesRule" and also check nonsense in the directive, like `'foo'` instead of 0/1.

File: src/Rules/Keywords/DeclareStrictTypesRule.php [hunk 1]
```diff
@@ -0,0 +1,69 @@
...
+	{
+		$declaresStrictTypes = false;
+		foreach ($node->declares as $declare) {
+			if (
+				$declare->key->name !== 'strict_types'
+			) {
+				continue;
+			}
+
+			$validValues = new UnionType([new ConstantIntegerType(0), new ConstantIntegerType(1)]);
```
Comment (ondrejmirtes) [inline src/Rules/Keywords/DeclareStrictTypesRule.php]: Please don't use the type system for this, just the AST. I'm pretty sure some stuff is not allowed that would slip through this.

**Review by ondrejmirtes — CHANGES_REQUESTED** [review]

File: src/Rules/Keywords/DeclareStrictTypesRule.php [hunk 2]
```diff
@@ -0,0 +1,70 @@
...
+				continue;
+			}
+
+			if (
+				!$declare->value instanceof Node\Scalar\LNumber
+				|| !in_array($declare->value->value, [0, 1], true)
+			) {
+				return [
+					RuleErrorBuilder::message(sprintf(
+						'Declare strict_types must have 0 or 1 as its value.',
```
Comment (ondrejmirtes) [inline src/Rules/Keywords/DeclareStrictTypesRule.php]: This message could be improved with `'Declare strict_types must have 0 or 1 as its value, %s given.'`

**Review by ondrejmirtes — CHANGES_REQUESTED** [review]

**Comment by ondrejmirtes** [comment]
Thank you!

**Comment by ondrejmirtes** [comment]
Hey @staabm this rule now causes a false positive: https://phpstan.org/r/adec71c6-2f0f-41b0-afc2-5df3f9110bce

Please look into it as it's currently blocking a release. Thank you.

**Comment by staabm** [comment]
I am on vacation and can have a look on sunday

**Pull request merged**